pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -622,11 +623,36 @@ func (ssc *defaultStatefulSetControl) rollingUpdateStatefulsetPods(
 		}
 	}
 
+	// handle specified deleted pod, and protected by maxUnavailable
+	for target := len(replicas) - 1; target >= 0; target-- {
+		if replicas[target] == nil {
+			continue
+		}
+		// the unavailable pods count exceed the maxUnavailable and the target is available, so we can't process it,
+		// why skip here rather than return?
+		// case: pod 0 ready, pod1 unready, pod 2 unready, pod3 ready, pod4 ready
+		// when maxUnavailable = 3, pod4 with specified deleted will be deleted but pod3 can't
+		// pod 2 and pod1 can be deleted because they were unavailable
+		if len(unavailablePods) >= maxUnavailable && !unavailablePods.Has(replicas[target].Name) {
+			continue
+		}
+		if specifieddelete.IsSpecifiedDelete(replicas[target]) {
```

Comment:
reverse the if clause and continue if not IsSpecifiedDelete so as to make the main logic more clear